qiskit/circuit/library/arithmetic/multipliers/hrs_cumulative_multiplier.py

```diff
@@ -0,0 +1,135 @@
+
+"""Compute the product of two qubit registers using classical multiplication approach."""
+
+from typing import Optional
+from qiskit.circuit import QuantumRegister, AncillaRegister, QuantumCircuit
+
+from .multiplier import Multiplier
+
+
+class HRSCumulativeMultiplier(Multiplier):
+    r"""A multiplication circuit to store product of two input registers out-of-place.
+
+    Circuit uses the approach from [1]. As an example, a multiplier circuit that
+    performs a non-modular multiplication on two 3-qubit sized registers with
+    the default adder is as follows:
+
+    .. parsed-literal::
+
+          a_0: ────■─────────────────────────
+                   │
+          a_1: ────┼─────────■───────────────
+                   │         │
+          a_2: ────┼─────────┼─────────■─────
+               ┌───┴────┐┌───┴────┐┌───┴────┐
+          b_0: ┤0       ├┤0       ├┤0       ├
+               │        ││        ││        │
+          b_1: ┤1       ├┤1       ├┤1       ├
+               │        ││        ││        │
+          b_2: ┤2       ├┤2       ├┤2       ├
+               │        ││        ││        │
+        out_0: ┤3       ├┤        ├┤        ├
+               │        ││        ││        │
+        out_1: ┤4       ├┤3       ├┤        ├
+               │  Adder ││  Adder ││  Adder │
+        out_2: ┤5       ├┤4       ├┤3       ├
+               │        ││        ││        │
+        out_3: ┤6       ├┤5       ├┤4       ├
+               │        ││        ││        │
+        out_4: ┤        ├┤6       ├┤5       ├
+               │        ││        ││        │
+        out_5: ┤        ├┤        ├┤6       ├
+               │        ││        ││        │
+        aux_0: ┤7       ├┤7       ├┤7       ├
+               └────────┘└────────┘└────────┘
+
+    Multiplication in this circuit is implemented in a classical approach by performing
+    a series of shifted additions using one of the input registers while the qubits
+    from the other input register act as control qubits for the adders.
+
+    **References:**
+
+    [1] Häner et al., Optimizing Quantum Circuits for Arithmetic, 2018.
+    `arXiv:1805.12445 <https://arxiv.org/pdf/1805.12445.pdf>`_
+
+    """
+
+    def __init__(
+        self,
+        num_state_qubits: int,
+        num_result_qubits: Optional[int] = None,
+        adder: Optional[QuantumCircuit] = None,
+        name: str = "HRSCumulativeMultiplier",
+    ) -> None:
+        r"""
+        Args:
+            num_state_qubits: The number of qubits in either input register for
+                state :math:`|a\rangle` or :math:`|b\rangle`. The two input
+                registers must have the same number of qubits.
+            num_result_qubits: The number of result qubits to limit the output to.
+                If number of result qubits is `n`, modulo :math:`2^n` is performed
+                to limit the output to the specified number of qubits. Default
+                value is ``2 * num_state_qubits``.
```

Comment:
```suggestion
                value is ``2 * num_state_qubits`` to represent any possible
                result from the multiplication of the two inputs.
```

Reply:
Thanks for the suggestion, a lot more understandable like this :)